### ISSUE TITLE
OIDC - API for admin to revoke user session

### ISSUE THREAD

**Issue by mgolding5010** [solo-io/gloo#10277]
### Gloo Edge Product

Enterprise

### Gloo Edge Version

1.17.2

### Is your feature request related to a problem? Please describe.

It is possible to revoke a user session in OP but accessToken may still be valid. At least when extAuth and redis is used, the redis key can also be deleted and the session is ended, forcing the user to log in again. Without Redis, client / browser can continue to gain access until the accessToken is invalid but extAuth does not have control of this.

### Describe the solution you'd like

Allow an Administrator to hit the logout endopoint for a user to log them out (i.e. logout of OP and remove redie key used by extAuth). NB: Admin needs mechanism to know which session key relates to which user. NB: Customer is not currently using gloo portal UI. Csutomer has requested an API for this. 

### Describe alternatives you've considered

Customer is prepared to write the functionality themself if solo aren't able to provide a solution.

**Comment by anessi** [comment]
Also see https://github.com/solo-io/gloo/issues/4446 for the initial request.

**Comment by DuncanDoyle** [comment]
I'm a bit concerned about the implications. Couple of questions and points of discussion:

1. For Gloo to be able to log a user out of their session in the OP, Gloo needs to have admin access to the OP. Giving a platform like Gloo admin access to an IdP/OP is not something that most security teams are keen on doing.
   - Gloo could use the user's access token that is stored in Redis, but that would mean the admin would somewhat "impersonate" the user by using their cached access-token to hit the logout endpoint.
 2. Given the above, if Gloo needs to use the OP's admin endpoint to revoke a user session, it needs to know the Admin API for the given OP. AFAIK there is no standard for OP admin APIs. E.g. Keycloak's admin endpoint to do this is `POST /admin/realms/{realm}/users/{user-id}/logout` (see: https://www.keycloak.org/docs-api/latest/rest-api/index.html), but I'm pretty sure that is not a standard.
 3. Different routes can use different AuthConfigs and thus point to different IdPs. AFAIK, we do not store the AuthConfig name and/or IdP config in Redis when we store the session. I'm therefore not sure how we would be able to correlate a user's session in Redis with a given OP configured in an AuthConfig.
 4. AFAIK, the only thing we store in Redis is the session key and the access-token and identity-token. We do not correlate this to any username whatsover, as that is the task of the IdP/OP.
 
I was wondering if we could use something like our new "OIDC Front Channel Logout" functionality to accomplish something like this, but I'm not sure if that functionality works when an admin revokes a given user's session. "Front Channel Logout" basically sends a logout command to all application bound to the user's session, and in the case of Gloo, this would remove the user session from Redis ....

**Comment by mgolding5010** [comment]
Yes, OIDC Front Channel Logout would be a good option, I guess it would
need gloo to allow impersonation of the user on behalf of the admin.

Kind regards,

Martin Golding
Senior Technical Account Manager | EMEA

**Comment by DuncanDoyle** [comment]
@anessi

If we can solve this with Back Channel Logout, I'm happy to prioritize that feature. We actually have an open ticket for that: https://github.com/k8sgateway/k8sgateway/issues/6913 which I have on the product backlog to complete the OIDC Logout support.
